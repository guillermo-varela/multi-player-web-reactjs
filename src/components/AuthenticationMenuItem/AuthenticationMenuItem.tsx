import React, { useState, useEffect, useRef } from 'react';
import { match } from 'react-router-dom';
import Nav from 'react-bootstrap/Nav';
import Navbar from 'react-bootstrap/Navbar';
import { LinkContainer } from 'react-router-bootstrap';
import Spinner from 'react-bootstrap/Spinner';
import Button from 'react-bootstrap/Button';
import Overlay from 'react-bootstrap/Overlay';
import Popover from 'react-bootstrap/Popover';
import { useDispatch } from 'react-redux';
import { FormattedMessage } from 'react-intl';
import useTypedSelector from '../../hooks/useTypedSelector';
import { signOutAction } from '../../actions/authenticationActions';
import { signIn } from '../../model/routes';

const AuthenticationMenuItem: React.FC<{ match: match }> = (props: { match: match }) => {
  const [showErrorSignOut, setErrorSignOut] = useState(false);
  const authenticationState = useTypedSelector(state => state.authentication);
  const dispatch = useDispatch();
  const signOutRef = useRef<HTMLButtonElement>(null);
  const { match } = props;

  useEffect(() => {
    setErrorSignOut(authenticationState.isErrorSignOut);
    if (authenticationState.isErrorSignOut) {
      setTimeout(() => setErrorSignOut(false), 5000);
    }
  }, [authenticationState.isErrorSignOut]);

  return (
    <Nav>
      {
        authenticationState.user ?
          <Navbar.Text>
            <span className="align-middle">
              {authenticationState.user.username.substr(0, authenticationState.user.username.indexOf('@'))}:&nbsp;
            </span>
            {authenticationState.isPendingSignOut ?
              <Spinner animation="border" size="sm" variant="light" />
              :
              <Button variant="link" className="menu-link" ref={signOutRef as any}
                onClick={() => { dispatch(signOutAction()) }}>
                <FormattedMessage id='menu.signOut' />
              </Button>
            }
          </Navbar.Text>
          :
          <LinkContainer to={`${match.url}${signIn}`}>
            <Nav.Link active={false}>
              <FormattedMessage id='menu.signIn' />
            </Nav.Link>
          </LinkContainer>
      }
      {
        signOutRef.current &&
        <Overlay
          placement="bottom"
          show={showErrorSignOut}
          target={signOutRef.current}>
          <Popover id="popover-sign-out">
            <Popover.Title as="h3">
              <FormattedMessage id='menu.signOut.error.title' />
            </Popover.Title>
            <Popover.Content>
              <FormattedMessage id='menu.signOut.error.message' />
            </Popover.Content>
          </Popover>
        </Overlay>
      }
    </Nav>
  );
}

export default AuthenticationMenuItem;
