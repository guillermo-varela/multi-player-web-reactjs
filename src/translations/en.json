{
  "component.about.code": "The source code for this project can be found at:",
  "component.about.codeSplit": "Route based code splitting",
  "component.about.csrf": "Use of custom HTTP headers in stateless/sessionless environments to prevent CSRF attacks",
  "component.about.forms.formik": "Forms in React using Formik",
  "component.about.hello": "Hi again :)",
  "component.about.internationalization": "Support for multiple languages (internationalization) using react-intl",
  "component.about.introduction": "This web application was developed to learn about:",
  "component.about.jwt": "Stateless/sessionless authentication on a server using \"Javascript Object Signing and Encryption\" (JOSE) and \"JSON Web Token\" (JWT)\"",
  "component.about.openAPI": "OpenAPI 3",
  "component.about.pixabay": "The multimedia content was fetched using the {link}.",
  "component.about.pixabay.name": "Pixabay API",
  "component.about.redux.typescript": "Redux using TypeScript",
  "component.about.rest.webflux": "REST services with reactive architecture (non-blocking) with Spring WebFlux and Spring Security",
  "component.error.message": "Ops! An error occurred...",
  "component.signin.button.sending": "Sending...",
  "component.signin.button.submit": "Sign In",
  "component.signin.credentials.hint": "Use user@test.com as email and password to test",
  "component.signin.empty.email": "Enter an email",
  "component.signin.incorrect.credentials": "Incorrect credentials",
  "component.signin.invalid.email": "Enter a valid email",
  "component.signin.invalid.password": "Enter a password",
  "component.signin.label.email": "Email",
  "component.signin.label.password": "Password",
  "menu.about": "About",
  "menu.home": "Home",
  "menu.images": "Images",
  "menu.more": "View More",
  "menu.sections": "Sections",
  "menu.signIn": "Sign In",
  "menu.signOut": "Sign Out",
  "menu.signOut.error.message": "Sign out failed",
  "menu.signOut.error.title": "Ops!",
  "menu.videos": "Videos",
  "multimedia.source": "Content from ",
  "multimedia.type.image": "Image",
  "multimedia.type.video": "Video",
  "nature.hello": "Hello!",
  "nature.introduction": "On this site you can find multimedia content about nature.",
  "nature.multimedia.endOfContent": "That's all for now...",
  "nature.title": "Nature",
  "root.loading": "Loading...",
  "root.locale.label": "Language",
  "root.nature.description": "Images and videos of flora and fauna.",
  "root.space.description": "Images and videos of the universe.",
  "space.hello": "Hello!",
  "space.images.introduction": "Multiple images related to outer space.",
  "space.introduction": "On this site you will find multimedia content about outer space.",
  "space.menu.title": "Space",
  "space.title": "Outer Space",
  "space.videos.introduction": "Mind-blowing videos from outer space."
}
