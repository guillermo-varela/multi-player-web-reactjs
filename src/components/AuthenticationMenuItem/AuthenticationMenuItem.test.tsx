import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { IntlProvider } from 'react-intl';
import configureMockStore, { MockStoreEnhanced } from 'redux-mock-store';
import { match, MemoryRouter } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import { mount, ReactWrapper } from 'enzyme';
import renderer from 'react-test-renderer';
import Nav from 'react-bootstrap/Nav';
import Navbar from 'react-bootstrap/Navbar';
import Overlay, { OverlayProps } from 'react-bootstrap/Overlay';
import Popover from 'react-bootstrap/Popover';
import { LinkContainer } from 'react-router-bootstrap';
import en from '../../translations/en.json';
import AuthenticationMenuItem from './AuthenticationMenuItem';

describe('<AuthenticationMenuItem />', () => {
  const mockStore = configureMockStore([thunk]);
  let store: MockStoreEnhanced<unknown, {}>;
  const usernameToTest = 'user-test';
  const emailToTest = `${usernameToTest}@test.com`;
  const match: match = {
    url: 'test',
    path: 'test',
    isExact: true,
    params: {}
  };

  it('renders <Nav />', () => {
    const wrapper = mount(buildComponent());
    expect(wrapper.find(Nav).length).toBe(1);
  });

  it('renders <LinkContainer /> to sign in when not authenticated', () => {
    const wrapper = mount(buildComponent());
    const linkContainer = wrapper.find(LinkContainer);
    expect(linkContainer.length).toBe(1);
    expect(linkContainer.props()).toHaveProperty('to', `${match.url}/signIn`);
  });

  it('renders username and <Button /> to sign out when authenticated', () => {
    const wrapper = mount(buildComponent(true));
    const navBarText = wrapper.children().find(Navbar.Text);
    expect(navBarText.find('Button[className="menu-link"]').length).toBe(1);

    expect(navBarText.text()).toContain(`${usernameToTest}:`);
  });

  it('renders error message when sign out fails', async () => {
    const wrapper = mount(buildComponent(true, true));
    let overlay: ReactWrapper<OverlayProps, any, React.Component<{}, {}, any>> | null = null;
    await act(async () => {
      overlay = wrapper.find(Overlay);
    });
    expect(overlay!.length).toBe(1);

    const popover = overlay!.find(Popover);
    expect(popover.length).toBe(1);
    expect(popover.prop('id')).toBe('popover-sign-out');
  });

  it('matches unauthenticated snapshot', () => {
    const tree = renderer.create(buildComponent()).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches authenticated snapshot', () => {
    const tree = renderer.create(buildComponent(true)).toJSON();
    expect(tree).toMatchSnapshot();
  });

  const buildComponent = (isAuthenticated: boolean = false, isErrorSignOut: boolean = false): JSX.Element => {
    const authentication = isAuthenticated ?
      {
        user: {
          username: emailToTest
        },
        isErrorSignOut: isErrorSignOut
      }
      : {};
    store = mockStore({ authentication: authentication });

    return (<Provider store={store}>
      <MemoryRouter>
        <IntlProvider locale='en' messages={en}>
          <AuthenticationMenuItem match={match} />
        </IntlProvider>
      </MemoryRouter>
    </Provider>);
  };
});
