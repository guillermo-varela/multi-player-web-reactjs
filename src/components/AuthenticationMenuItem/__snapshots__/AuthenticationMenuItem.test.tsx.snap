// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthenticationMenuItem /> matches authenticated snapshot 1`] = `
<div
  className="nav"
  onKeyDown={[Function]}
>
  <span
    className="navbar-text"
  >
    <span
      className="align-middle"
    >
      user-test
      :Â 
    </span>
    <button
      className="menu-link btn btn-link"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Sign Out
    </button>
  </span>
</div>
`;

exports[`<AuthenticationMenuItem /> matches unauthenticated snapshot 1`] = `
<div
  className="nav"
  onKeyDown={[Function]}
>
  <a
    className="nav-link"
    data-rb-event-key="test/signIn"
    href="test/signIn"
    onClick={[Function]}
    onKeyDown={[Function]}
  >
    Sign In
  </a>
</div>
`;
